src/XrdCeph/XrdCephBulkAioRead.cc

```diff
@@ -0,0 +1,237 @@
+    } catch (std::bad_alloc&) {
+      log_func((char*)"Can not allocate int for read object op (%lu, %lu)", offset, size);
+      delete bl;
+      delete retval;
+      return -ENOMEM;
+    }
+
+    librados::AioCompletion *cmpl = librados::Rados::aio_create_completion();
+    if (0 == cmpl) {
+      log_func((char*)"Can not create completion for read (%lu, %lu)", offset, size);
+      delete bl;
+      delete retval;
+      delete rop;
+      return -ENOMEM;
+    }
+    operations.insert( {obj_idx, std::make_pair(rop, cmpl)} );
+  } else {
+    rop = std::get<0>(op_data->second);
+  }
+  buffers.push_back( std::make_tuple(bl, out_buf, retval) );
+
+  rop->read(offset, size, bl, retval);
+  return 0;
+}
+
+int bulkAioRead::submit_and_wait_for_complete() {
+  /**
+   * Submit previously prepared read requests and wait for their completion
+   *
+   * To prepare read requests use `read` or `addRequest` methods.
+   *
+   * @return  zero on success, negative error code on failure
+   *
+   */
+
+  std::string obj_name;
+  size_t obj_idx;
+  librados::AioCompletion* cmpl;
+  librados::ObjectReadOperation* op;
+
+
+  for (auto const& op_data: operations) {
+    obj_idx = op_data.first;
+    //16 bytes for object hex number, 1 for dot and 1 for null-terminator
+    char object_suffix[18];
+    int sp_bytes_written;
+    sp_bytes_written = snprintf(object_suffix, sizeof(object_suffix), ".%016zx", obj_idx);
+    if (sp_bytes_written > (int) sizeof(object_suffix)) {
+      log_func((char*)"Can not fit object suffix into buffer for file %s -- too big\n", file_name.c_str());
+      return -EFBIG;
+    } 
+
+    try {
+      obj_name =  file_name + std::string(object_suffix);
+    } catch (std::bad_alloc&) {
+      log_func((char*)"Can not create object string for file %s)", file_name.c_str());
+      return -ENOMEM;
+    }
+    op = std::get<0>(op_data.second);
+    cmpl = std::get<1>(op_data.second);
+    context->aio_operate(obj_name, cmpl, op, 0);
+  }
+
+  for (auto const& op_data: operations) {
+    cmpl = std::get<1>(op_data.second);
+    cmpl->wait_for_complete();
+  }
+  return 0;
+}
+
+ssize_t bulkAioRead::get_results() {
+  /**
+   * Copy the results of executed read requests from ceph's bufferlists to client's buffers
+   *
+   * Note that this method should be called only after the submission and completion of read
+   * requests, i.e. after `submit_and_wait_for_complete` method.
+   *
+   * @return  cumulative number of bytes read (by all read operations) on success, negative
+   *          error code on failure
+   *
+   */
+
+  ceph::bufferlist* bl;
+  char *buf;
+  int *rc;
+  ssize_t res = 0;
+  for (ReadOpData i: buffers) {
+    bl = std::get<0>(i);
+    buf = std::get<1>(i);
+    rc = std::get<2>(i);
+    if (*rc < 0) {
+      //Is it possible to get here?
+      log_func((char*)"One of the reads failed with rc %d", rc);
+      return *rc;
+    }
+    bl->begin().copy(bl->length(), buf);
+    res += bl->length();
```

Comment:
If `i` is the last buffer, and it presents fewer bytes than requested, you can treat this as a over-read of the tail of the file, and return a sum smaller than requested.  If `i` is not the last buffer, and there are fewer bytes, is there a way to signal to the client that there's a gap in the result?  If not, it would have to be an over-all error - that would be a limitation of the interface.
In the separate case of a sparse file, I'd expect `*rc` to indicate this, and the reaction would be to `memset(buf, 0, len)`.  But it's surprising if `*rc` is not indicating this, as there doesn't seem to be any other opportunity for librados to express this.
Either way, short of checking file lengths earlier, or checking for existence of objects, this seems the place to detect these conditions.

Reply:
Actually, unless you know the file length, you're not going to be able to distinguish between a sparse-missing object and a beyond-EOF-missing object.

Reply:
This seems to be not relevant since we are now relying on striper for sparse reads.